File: frontend/src/lib/conversations.ts
import { api } from "./api";

export type Sender = "client" | "agent" | "staff";
export type ConversationStatus = "attention" | "held" | "auto_sent";
export type ConversationFilter = "all" | "needs_attention" | "auto_sent";

export type Client = { name: string | null; phone_number: string };

export type LastMessage = { content: string; sender: Sender; created_at: string };

export type ConversationSummary = {
  id: string;
  channel: "whatsapp";
  client: Client;
  last_message: LastMessage | null;
  status: ConversationStatus;
  total: number | null;
};

export type Message = { id: string; sender: Sender; content: string; created_at: string };

export type PriceLine = {
  item_id: string;
  name: string;
  unit: string;
  quantity: number;
  unit_price: number;
  applied_min_qty: number;
  total: number;
};

export type QuoteSummary = {
  status: "pending" | "approved" | "rejected" | "auto_sent";
  lines: PriceLine[];
  subtotal: number;
};

export type ConversationDetail = {
  id: string;
  channel: "whatsapp";
  client: Client;
  messages: Message[];
  latest_quote: QuoteSummary | null;
};

export const conversationsApi = {
  list: (filter: ConversationFilter) =>
    api.get<ConversationSummary[]>(
      filter === "all" ? "/api/v1/conversations" : `/api/v1/conversations?status=${filter}`,
    ),
  get: (id: string) => api.get<ConversationDetail>(`/api/v1/conversations/${id}`),
};

File: app/routers/conversations.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.users import current_active_business
from app.db.session import get_session
from app.models.business import Business
from app.schemas.conversation import ConversationDetail, ConversationSummary, QuoteSummary
from app.schemas.pricing import PriceBreakdown
from app.services.conversation import (
    ConversationSummaryRow,
    derive_conversation_status,
    get_conversation_thread,
    list_conversations,
)

router = APIRouter()


def _quote_total(lines: list[dict] | None) -> int | None:
    if not lines:
        return None
    return sum(line["total"] for line in lines)


def _to_summary(row: ConversationSummaryRow) -> ConversationSummary:
    return ConversationSummary(
        id=row.conversation.id,
        channel=row.conversation.channel,
        client=row.client,
        last_message=row.last_message,
        status=derive_conversation_status(row.latest_quote),
        total=_quote_total(row.latest_quote.lines if row.latest_quote else None),
    )


# Filter tabs collapse the 3-way badge into 2 buckets — "needs attention"
# covers both held and attention, matching the design's two non-"all" tabs.
_STATUS_FILTERS = {
    "needs_attention": {"attention", "held"},
    "auto_sent": {"auto_sent"},
}


@router.get("", response_model=list[ConversationSummary])
async def list_conversations_endpoint(
    status_filter: str | None = Query(default=None, alias="status"),
    business: Business = Depends(current_active_business),
    session: AsyncSession = Depends(get_session),
) -> list[ConversationSummary]:
    rows = await list_conversations(session, business.id)
    summaries = [_to_summary(row) for row in rows]
    allowed = _STATUS_FILTERS.get(status_filter) if status_filter else None
    if allowed is not None:
        summaries = [s for s in summaries if s.status in allowed]
    return summaries


@router.get("/{conversation_id}", response_model=ConversationDetail)
async def get_conversation_endpoint(
    conversation_id: uuid.UUID,
    business: Business = Depends(current_active_business),
    session: AsyncSession = Depends(get_session),
) -> ConversationDetail:
    thread = await get_conversation_thread(session, business.id, conversation_id)
    if thread is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND)

    latest_quote = None
    if thread.latest_quote is not None and thread.latest_quote.lines:
        latest_quote = QuoteSummary(
            status=thread.latest_quote.status,
            lines=[PriceBreakdown.model_validate(line) for line in thread.latest_quote.lines],
            subtotal=_quote_total(thread.latest_quote.lines),
        )

    return ConversationDetail(
        id=thread.conversation.id,
        channel=thread.conversation.channel,
        client=thread.client,
        messages=thread.messages,
        latest_quote=latest_quote,
    )
